DRAFT.md

```diff
@@ -8,30 +8,141 @@ Working Group: Library, Core, Evolution
+The same applies to Win32 asynchronous callbacks.
+
+The  IEEE  Std  1003.1-2004  ("POSIX.1”) revision marked the functions makecontext() and swapcontext() (the ucontext APIs) as obsolete, noting that they were intended for use for safely returning from a signal handler into the signalled frame (which is not possible with setjmp / longjmp) and were not intended to be used for threading.
+The IEEE Std 1003.1-2008 ("POSIX.1") revision removed the functions from the specification.
+
+Correctly implementing this would require code that is specific to both architectures and operating systems.
+This would make it unique in the C++ standard library.
+Some features are architecture-specific but operating-system agnostic, some are operating-system specific but architecture-agnostic.
+No existing features have this property.
+
+### Users of TLS expect it to be private
+
+Thread-local state is expected to be modified only in the same lexical scopes.
+P0876 explicitly notes this in one context: exceptions.
+The exception implementation typically maintains a list of in-flight exceptions in thread-local storage.
+If a fiber switch occurs in a `catch` block, this linked list can be corrupted.
+
+Unfortunately, this is a general problem.
+If a library creates any kind of linked list in thread-local storage that follows lexical scope, this will be corrupted.
+Variations of this idiom occur in various forms.
+For example:
+
+ - Using a parallel stack for large returns.
+ - Managing in-flight transactions for software-transactional memory.
+ - Providing exception-like behaviour with `setjmp` for languages without exceptions.
+
+These and more may exist in libraries called by C++.
+If any such library calls a C++ function that yields, it may resume with this state corrupted.
+
+Coroutines do not have this problem because the point at which a coroutine is created and the point at which it yields are both visible.
+A coroutine cannot yield with library code on the stack.
+
+### Accidental deadlocks
+
+Systems that provide a scheduler and N:M threading modify their locks to make them fiber-aware.
+This is essential to avoid accidental deadlocks.
+For example, consider the simple case of a library that implements thread safety with a global lock.
+Each function will look something like this:
 
-### Sample code that illustrates the issue
+```c++
+std::mutex giant;
 
-```cpp
-TODO
+void someFunction()
+{
+    std::lock_guard g{giant};
+    ...
+}
 ```
 
-## References
+This trivial case is safe, but consider this small change:
+
+
+```c++
+struct UserExtensibleType
+{
+    virtual void some_method() = 0;
+};
+
+std::mutex giant;
+
+void someFunction(UserExtensibleType &object)
+{
+    std::lock_guard g{giant};
+    ...
+    object.some_method();
+}
+```
+
+If the user's implementation of `some_method` yields (directly or indirectly) and the new fiber invokes another function from this library, the thread deadlocks.
```

Comment:
I haven't digested P0876R19 in enough detail, but is this _guaranteed_ to deadlock under only the assumption listed, or only under some conditions, i.e., if there is a second fibre to yield which itself will never yield and never returns? Perhaps p21 prevents this in theory ("Every `fiber_context` you launch must terminate gracefully by returning from its entry-function.")

Reply:
> but is this guaranteed to deadlock under only the assumption listed,

Yes.  Unless the lock is modified to yield to the owning fibre.

> if there is a second fibre to yield which itself will never yield and never returns

Acquiring the lock will not yield unless the lock is modified to yield, and system locks are not (see the layering problems section).

> Perhaps p21 prevents this in theory ("Every fiber_context you launch must terminate gracefully by returning from its entry-function.")

I presume this actually means 'if a context terminates, it must do so by returning to the function that created it', rather than 'you must solve the halting problem over arbitrary third-party code to use fibres'?